### PR TITLE
feat: test dag order framework

### LINKED ISSUE #299: [test] Test reconfigure order
We should provide a generic way to test that reconfiguring a service will trigger the necessary services.

We should be able, for instance, to make sure that reconfiguring hdfs will restart hbase.

### PR DESCRIPTION
**Pull request by PaulFarault**
#### Which issue(s) this PR fixes

Fixes #299

This PR provide code to generate tests that will check if the DAG order is correct based on rules specified in each collections.

Usage instructions have been defined in a `README.md`.

#### Additional comments

I'm not sure about the packaging of those tests as:

- They must be run from the directory containing the `ansible.cfg` file.
- `pytest` and `tdp-lib` must be available.

#### Agreements

- [X] I hereby declare this contribution to be licensed under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0).
- [X] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

**Comment by PaulFarault** [comment]
Nodes are currently displayed in alphabetical order, for convenience. Would it make more sense to leave them in topological order?

File: test_dag_order/helpers.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+            with open(rules_file_path) as f:
+                rules: dict = yaml.safe_load(f)
+                for [source, rule_dict] in rules.items():
+                    yield _Rule.from_rule_dict(
+                        source=source, rules=rule_dict, rules_file_path=rules_file_path
+                    )
+
+
+class append_collection_list_action(argparse.Action):
+    """argpase action to create a list of CollectionToTest from paths"""
```
Comment (rpignolet) [inline test_dag_order/helpers.py:168]:
```suggestion
    """argparse action to create a list of CollectionToTest from paths"""
```

File: test_dag_order/helpers.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+    Returns:
+        Tuple of the resolved components and the service component map.
+
+    Examples:
+        >>> resolve_components(["service1", "service2"])
+        (frozenset({"service1", "service2"}), {})
+        >>> resolve_components(["service1", "service2_component1"])
+        (frozenset({"service1", "service2_component1"}), {"service2_component1": "service2"})
+        >>> resolve_components(["service1", "service2"])
```
Comment (rpignolet) [inline test_dag_order/helpers.py:206]: The third example is the same as the first and the fourth example is the same as the second right?

File: test_dag_order/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+Every `.yml` file placed in the `tdp_lib_dag_test_rules` of a TDP collection can contain a list of rules to test. Each rule is a dictionary where:
+
+- the key is the `source` component (or service) to reconfigure.
+- the properties are:
+  - `must_include`: a list of components (or services) that must be reconfigured
+  - `must_exclude`: a list of components (or services) that must not be reconfigured
+
+
+For example, the following rule will test that editing the `hdfs.yml` configuration will trigger a reconfigure that must include the `yarn` service and must exclude the `zookeeper` service:
```
Comment (rpignolet) [inline test_dag_order/README.md:37]: `must_include: ["yarn"]` is used to include all `yarn` components right?
```suggestion
For example, the following rule will test that editing the `hdfs.yml` configuration will trigger a reconfigure that must include all components of `yarn` service and must exclude all components of `zookeeper` service:
```

**Comment by PaulFarault** [comment]
@rpignolet I fixed everything. Should we move it in it's own repo?

**Comment by PaulFarault** [comment]
Force pushed after rebase to include #532

**Review by rpignolet — APPROVED** [review]

**Pull request merged**